libvast/src/system/explorer.cpp

```diff
@@ -41,43 +41,36 @@ explorer_state::explorer_state(caf::event_based_actor*) {
 }
 
 void explorer_state::forward_results(vast::table_slice slice) {
+  // Discard all if the client already has enough.
+  if (num_sent == limits.total)
+    return;
   // Check which of the ids in this slice were already sent to the sink
   // and forward those that were not.
-  vast::bitmap unseen;
+  ewah_bitmap unseen = {};
+  const auto offset = slice.offset() == invalid_id ? 0 : slice.offset();
+  unseen.append_bits(false, offset);
   for (size_t i = 0; i < slice.rows(); ++i) {
-    auto id = slice.offset() + i;
-    auto [_, new_] = returned_ids.insert(id);
-    if (new_) {
-      vast::ids tmp;
-      tmp.append_bits(false, id);
-      tmp.append_bits(true, 1);
-      unseen |= tmp;
-    }
+    default_hash h;
+    for (size_t j = 0; j < slice.columns(); ++j)
+      hash_append(h, slice.at(i, j));
+    auto digest = h.finish();
+    auto [_, new_] = returned_ids.insert(digest);
+    unseen.append_bits(new_, 1);
+    if (num_sent + rank(unseen) == limits.total)
+      break;
   }
```

Comment:
Note that the summarize plugin already does the same thing for a subset of columns (those that we group by) but in a columnar fashion. Should we do that here as well?

Reply:
Do you mean creating a "column" of hashers and moving them over the slice left to right?

Reply:
Yes, and directly on the record batch as well. Pretty sure this isn't hard to write, the question is rather whether we want to spend the hour here.

Reply:
Imho we don't.